### PR TITLE
feat: roster change objects

### LINKED ISSUE #9907: Create objects for communicating a change in roster
Create the objects that will be used for communicating a change in roster.

### PR DESCRIPTION
**Pull request by cody-littley**
Closes #9907

### CODE REVIEW

**Review by alittley — APPROVED** [review]

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/address/AddressBook.java [hunk 1]
```diff
@@ -178,7 +178,7 @@ public long getClassId() {
      * @return the round when this address book was constructed, or {@link #UNKNOWN_ROUND} if the round is unknown or
      * this address book was not constructed during a regular round
      */
-    public long getRound() {
+    public long getRound() { // TODO remove
```
Comment (edward-swirldslabs) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/address/AddressBook.java]: Is this meant to be future work?  Is the method deprecated?

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterDiffGenerator.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+     */
+    @Nullable
+    public RosterDiff generateDiff(@NonNull final UpdatedRoster updatedRoster) {
+        final AddressBook roster = updatedRoster.roster();
+        final long effectiveRound = updatedRoster.effectiveRound();
+
+        try {
+            if (previousRoster == null) {
+                // This is the first roster we've seen.
+                return null;
```
Comment (edward-swirldslabs) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterDiffGenerator.java:69]: When I see a return, I want to know that the private data fields are determined.  Wrapping return statements in `try-finally` blocks means I have to know the impacts of everything in the finally block and it is easy to miss they exist.  (A linear reading of the code is not possible, I have to know ahead of time that I need to jump to the finally block to get the complete picture.)    Personally I think this pattern is a code-smell.   It saves lines of code, but makes the code more complicated to review.

Comment (edward-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterDiffGenerator.java:69]: In this case I was going to observe that the `previousRoster` and `previousEffectiveRound` were not set before returning.   But then I realized the finally was doing that work.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterDiffGenerator.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+                        "Effective rounds should always increase over time. "
+                                + "Previous effective round: {}, new effective round: {}",
+                        previousEffectiveRound,
+                        effectiveRound);
+            }
+
+            if (roster.getHash() == null) {
+                // FUTURE WORK: when this is wired into the system, make sure this is thread safe
+                cryptography.digestSync(roster);
+            }
```
[the remark this answers is not in the document]
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterDiffGenerator.java:84]: I agree.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterDiffGenerator.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+import java.util.Set;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Takes as input a sequence of {@link UpdatedRoster} objects and produces a sequence of {@link RosterDiff} objects.
+ */
+public class RosterDiffGenerator {
+
+    private static final Logger logger = LogManager.getLogger();
```
[the remark this answers is not in the document]
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterDiffGenerator.java:39]: Oversight, we should always specify class. Added.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterDiff.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ * @param membershipChanged      whether the membership changed
+ * @param addedNodes             the nodes that were added
+ * @param removedNodes           the nodes that were removed
+ * @param modifiedNodes          the nodes that were modified
+ */
+public record RosterDiff(
+        @NonNull UpdatedRoster newRoster,
+        boolean rosterIsIdentical,
+        boolean consensusWeightChanged,
+        boolean membershipChanged,
```
[the remark this answers is not in the document]
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterDiff.java:38]: We can deduce everything in this record. The idea is that we do all of the comparison in one location, and then downstream classes can take a shortcut and just look at these values.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterDiffGenerator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterDiffGenerator.java:84]: Removed this hash, will ensure it is hashed when this is wired in

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/address/AddressBook.java (hunk 1 above)
Comment (cody-littley) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/address/AddressBook.java]: Should not be done in this PR. TODO has been removed.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterDiffGenerator.java (hunk 2 above)
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterDiffGenerator.java:69]: refactored to be linear

**Review by alittley — APPROVED** [review]

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterDiffGenerator.java (hunk 2 above)
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterDiffGenerator.java]: much appreciated

**Pull request merged**